core/src/main/scala/org/apache/spark/util/CheckpointingIterator.scala

```diff
@@ -0,0 +1,225 @@
+ * @param tempOutputPath the temp path to write the checkpoint data
+ * @param finalOutputPath the final path to move the temp file to when finishing checkpointing
+ * @param context the task context
+ * @param blockSize the block size for writing the checkpoint data
+ */
+private[spark] class CheckpointingIterator[T: ClassTag](
+    id: RDDBlockId,
+    values: Iterator[T],
+    fs: FileSystem,
+    tempOutputPath: Path,
+    finalOutputPath: Path,
+    context: TaskContext,
+    blockSize: Int) extends Iterator[T] with Logging {
+
+  private[this] var completed = false
+
+  // We don't know if the task is successful. So it's possible that we still checkpoint the
+  // remaining values even if the task is failed.
+  // TODO optimize the failure case if we can know the task status
+  context.addTaskCompletionListener { _ => complete() }
+
+  private[this] val fileOutputStream: OutputStream = {
+    val bufferSize = SparkEnv.get.conf.getInt("spark.buffer.size", 65536)
+    if (blockSize < 0) {
+      fs.create(tempOutputPath, false, bufferSize)
+    } else {
+      // This is mainly for testing purpose
+      fs.create(tempOutputPath, false, bufferSize, fs.getDefaultReplication, blockSize)
+    }
+  }
+
+  private[this] val serializeStream =
+    SparkEnv.get.serializer.newInstance().serializeStream(fileOutputStream)
+
+  /**
+   * Called when this iterator is on the last element by `hasNext`.
+   * This method will rename temporary output path to final output path of checkpoint data.
+   */
+  private[this] def complete(): Unit = {
+    if (completed) {
+      return
+    }
+
+    if (serializeStream == null) {
+      // There is some exception when creating serializeStream, we only need to clean up the
+      // resources.
+      cleanup()
+      return
+    }
+
+    while (values.hasNext) {
+      serializeStream.writeObject(values.next)
+    }
+
+    completed = true
+    CheckpointingIterator.releaseLockForPartition(id)
+    serializeStream.close()
+
+    if (!fs.rename(tempOutputPath, finalOutputPath)) {
+      if (!fs.exists(finalOutputPath)) {
+        logInfo("Deleting tempOutputPath " + tempOutputPath)
+        fs.delete(tempOutputPath, false)
+        throw new IOException("Checkpoint failed: failed to save output of task: "
+          + context.attemptNumber + " and final output path does not exist")
+      } else {
+        // Some other copy of this task must've finished before us and renamed it
+        logInfo("Final output path " + finalOutputPath + " already exists; not overwriting it")
+        fs.delete(tempOutputPath, false)
+      }
+    }
+  }
+
+  private[this] def cleanup(): Unit = {
+    completed = true
+    CheckpointingIterator.releaseLockForPartition(id)
+    if (serializeStream != null) {
+      serializeStream.close()
+    }
+    fs.delete(tempOutputPath, false)
+  }
+
+  private[this] def cleanupOnFailure[A](body: => A): A = {
+    try {
+      body
+    } catch {
+      case e: Throwable =>
```

Comment:
Use `Throwable` here because it will be thrown later. It's better to cleanup as well for fatal errors.